flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/JoinedRow.java

```diff
@@ -0,0 +1,303 @@
+			return row1.getGeneric(i);
+		} else {
+			return row2.getGeneric(i - row1.getArity());
+		}
+	}
+
+	@Override
+	public BinaryString getString(int i) {
+		if (i < row1.getArity()) {
+			return row1.getString(i);
+		} else {
+			return row2.getString(i - row1.getArity());
+		}
+	}
+
+	@Override
+	public BaseRow getRow(int i, int numFields) {
+		if (i < row1.getArity()) {
+			return row1.getRow(i, numFields);
+		} else {
+			return row2.getRow(i - row1.getArity(), numFields);
+		}
+	}
+
+	@Override
+	public BinaryArray getArray(int i) {
+		if (i < row1.getArity()) {
+			return row1.getArray(i);
+		} else {
+			return row2.getArray(i - row1.getArity());
+		}
+	}
+
+	@Override
+	public BinaryMap getMap(int i) {
+		if (i < row1.getArity()) {
+			return row1.getMap(i);
+		} else {
+			return row2.getMap(i - row1.getArity());
+		}
+	}
+
+	@Override
+	public void setNullAt(int i) {
+		if (i < row1.getArity()) {
+			row1.setNullAt(i);
+		} else {
+			row2.setNullAt(i - row1.getArity());
+		}
+	}
+
+	@Override
+	public void setBoolean(int i, boolean value) {
+		if (i < row1.getArity()) {
+			row1.setBoolean(i, value);
+		} else {
+			row2.setBoolean(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setByte(int i, byte value) {
+		if (i < row1.getArity()) {
+			row1.setByte(i, value);
+		} else {
+			row2.setByte(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setShort(int i, short value) {
+		if (i < row1.getArity()) {
+			row1.setShort(i, value);
+		} else {
+			row2.setShort(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setInt(int i, int value) {
+		if (i < row1.getArity()) {
+			row1.setInt(i, value);
+		} else {
+			row2.setInt(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setLong(int i, long value) {
+		if (i < row1.getArity()) {
+			row1.setLong(i, value);
+		} else {
+			row2.setLong(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setFloat(int i, float value) {
+		if (i < row1.getArity()) {
+			row1.setFloat(i, value);
+		} else {
+			row2.setFloat(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setDouble(int i, double value) {
+		if (i < row1.getArity()) {
+			row1.setDouble(i, value);
+		} else {
+			row2.setDouble(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setChar(int i, char value) {
+		if (i < row1.getArity()) {
+			row1.setChar(i, value);
+		} else {
+			row2.setChar(i - row1.getArity(), value);
+		}
+	}
+
+	@Override
+	public void setDecimal(int i, Decimal value, int precision) {
+		if (i < row1.getArity()) {
+			row1.setDecimal(i, value, precision);
+		} else {
+			row2.setDecimal(i - row1.getArity(), value, precision);
+		}
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || !(o instanceof JoinedRow)) {
+			return false;
+		}
+
+		JoinedRow row = (JoinedRow) o;
+
+		return this.header == row.header &&
+			row1.equals(row.row1) &&
```

Comment:
I still think that we should ignore row1 and row2's header when comparing. The only meaningful header is `JoinedRow`'s, row1 and row2 only contribute data

Reply:
setHeader of row1 and row2 while setHeader of this JoinedRow?

Reply:
We should throw Exception when equals and hashcode in joinedRow.